pkg/webhook/client.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+Copyright 2024 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// IsPreprovisionCSIVolume checks whether the volume is a pre-provisioned volume for the desired csiDriver.
+func (si *SidecarInjector) IsPreprovisionCSIVolume(csiDriver string, pvc *corev1.PersistentVolumeClaim) (bool, error) {
+	pv, err := si.GetPreprovisionCSIVolume(csiDriver, pvc)
+	if pv != nil {
+		return true, err
+	}
+
+	return false, err
+}
+
+// GetPreprovisionCSIVolume gets the pre-provisioned persistentVolume when backed by the desired csiDriver.
+func (si *SidecarInjector) GetPreprovisionCSIVolume(csiDriver string, pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolume, error) {
+	if csiDriver == "" {
+		return nil, errors.New("csiDriver is empty, cannot verify storage type")
+	}
+
+	if pvc == nil {
+		return nil, errors.New("pvc is nil, cannot get pv")
+	}
+
+	if pvc.Spec.VolumeName == "" {
+		return nil, errors.New("pvc.spec.volumeName is not set, cannot find pv")
+	}
+
+	// GetPV returns an error if pvc.Spec.VolumeName is not empty and the associated PV object is not found in the API server.
+	pv, err := si.GetPV(pvc.Spec.VolumeName)
+	if err != nil {
+		return nil, err // no additional context needed for error.
+	}
+
+	if pv.Spec.CSI == nil {
+		return nil, fmt.Errorf("perisistentVolume is not backed by %s", csiDriver)
+	}
+
+	// PV is using dynamic mounting, which does not count as a preprovisioned volume.
+	// See details: https://cloud.google.com/storage/docs/gcsfuse-mount#dynamic-mount
+	if csiDriver == gcsFuseCsiDriverName && pv.Spec.CSI.VolumeHandle == "_" {
```

Comment:
I don't understand this logic, why are we returning error if volumeHandle is "_". Can you please elaborate? 
Are we saying that for dynamic mounting PV/PVC way of expressing that intent is not supported?

Reply:
This function should only check if its a persistent volume. Moved this check to `isGcsFuseVolume`